ml-agents/mlagents/trainers/torch/layers.py

```diff
@@ -323,27 +324,26 @@ def __init__(
             layer_in_size = layer_size
         flat_output = linear_layer(
             layer_size,
-            input_size * output_size + output_size,
+            input_size * output_size,
             kernel_init=Initialization.KaimingHeNormal,
             kernel_gain=0.1,
             bias_init=Initialization.Zero,
         )
-        self.hypernet = torch.nn.Sequential(*layers, flat_output)
 
-    def forward(self, input_activation, hyper_input):
-        flat_output_weights = self.hypernet(hyper_input)
-        batch_size = input_activation.size(0)
+        # Re-initializing the weights of the last layer of the hypernetwork
+        bound = math.sqrt(1 / (layer_size * self.input_size))
+        flat_output.weight.data.uniform_(-bound, bound)
 
-        output_weights, output_bias = torch.split(
-            flat_output_weights, self.input_size * self.output_size, dim=-1
-        )
+        self.hypernet = torch.nn.Sequential(*layers, LayerNorm(), flat_output)
 
-        output_weights = output_weights.view(
-            batch_size, self.input_size, self.output_size
-        )
-        output_bias = output_bias.view(batch_size, self.output_size)
-        output = (
+        # The hypernetwork will not generate the bias of the main network layer
+        self.bias = torch.nn.Parameter(torch.zeros(output_size))
```

Comment:
Did you find that performance was worse when generating the biases also?

Reply:
Yes, it worked better with fixed biases. But I think it is because we use the same initialization for both weights and biases. Maybe we will see improvements if we have a different path for generating the biases.